core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala

```diff
@@ -478,20 +633,55 @@ final class DefaultSource extends ColumnarRelationProvider {
       schema, externalStore)
 
     var success = false
-    val relation = new ColumnFormatRelation(SnappyStoreHiveCatalog.
+
+    // create an index relation if it is an index table
+    val relation = options.get(StoreUtils.GEM_INDEXED_TABLE) match {
+      case Some(baseTable) => new IndexColumnFormatRelation(SnappyStoreHiveCatalog.
+        processTableIdentifier(table, sqlContext.conf),
+        getClass.getCanonicalName,
+        mode,
+        schema,
+        schemaExtension,
+        ddlExtensionForShadowTable,
+        options,
+        externalStore,
+        blockMap,
+        partitioningColumn,
+        sqlContext,
+        baseTable)
+      case None => new ColumnFormatRelation(SnappyStoreHiveCatalog.
         processTableIdentifier(table, sqlContext.conf),
-      getClass.getCanonicalName,
-      mode,
-      schema,
-      schemaExtension,
-      ddlExtensionForShadowTable,
-      options,
-      externalStore,
-      blockMap,
-      partitioningColumn,
-      sqlContext)
+        getClass.getCanonicalName,
+        mode,
+        schema,
+        schemaExtension,
+        ddlExtensionForShadowTable,
+        options,
+        externalStore,
+        blockMap,
+        partitioningColumn,
+        sqlContext)
+    }
+
     try {
       relation.createTable(mode)
+      if (relation.isInstanceOf[ColumnFormatRelation]) {
+        var indexes: Array[String] = Array()
+        try {
+          val options = new CaseInsensitiveMutableHashMap(relation.origOptions)
+          indexes = options(ExternalStoreUtils.INDEX_NAME).split(",")
+        } catch {
+          case e: NoSuchElementException =>
+        }
+
+        indexes.foreach(index => {
+          val sncCatalog = sqlContext.asInstanceOf[SnappyContext].catalog
+          val dr = sncCatalog.lookupRelation(sncCatalog.newQualifiedTableName(index)) match {
+            case LogicalRelation(r: DependentRelation, _) => r
+          }
+          relation.asInstanceOf[ColumnFormatRelation].addDependent(dr, sncCatalog)
+        })
+      }
```

Comment:
Ok. I created a new function recoverDependentsRelation that recreates any DependentRelation and add them as dependent. So I have folded the two step strategy you have suggested in single step. However, I have added a new ExternalTableType for indexes which is not getting used now but can be used in future.